odl/discr/grid.py

```diff
@@ -226,12 +226,22 @@ def coord_vectors(self):
     @property
     def ndim(self):
         """Number of dimensions of the grid."""
-        return len(self.coord_vectors)
+        try:
+            return self.__ndim
+        except AttributeError:
+            ndim = len(self.coord_vectors)
+            self.__ndim = ndim
```

Comment:
As I had written somewhere else, I'm not very fond of the `try: self.bla except AttributeError: self.bla = ...` pattern. Setting the attributes in question to `None` in the initializer together with a comment makes the intent much clearer and makes cached attributes easier to find.

Reply:
I just stumbled over this while compiling the release notes.

What you wrote @adler-j may have been true for older versions of Python. Since 3.6, however, dictionaries have an [internal version](https://www.python.org/dev/peps/pep-0509/) (in the C implementation) that lets the interpreter see if the dictionary has changed since the last access. If not, a cached version is used, several times faster than the full lookup.

The effect of this is that classes that define all of their instance attributes in `__init__` get rewarded by faster lookups later on, since the instance `__dict__` stays constant.

Anyway, we should revise this pattern and do some benchmarking to get the numbers.

Reply:
I agree that we should try.

With that said, wouldnt the cache update after the first access anyway, evening out in the long term?

Reply:
Yeah, I'm not entirely sure what the implications are. I just heard one of the core Python devs mention this detail as a reason to set up all attributes during `__init__`, but you're right, it might just play a role for the first lookup. Anyway, should be easy to test.